### PR TITLE
Upgrade torch nightly version to 2.8.0.dev20250408+cpu

### PR DESCRIPTION
**Pull request by dayo09**
Let's upgrade torch nightly version to the latest.

TICO-DCO-1.0-Signed-off-by: Dayoung Lee <dayoung.lee@samsung.com>

---
Related to: #9

### CODE REVIEW

File: .github/workflows/check-pr.yaml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
       image: ubuntu:22.04
     strategy:
       matrix:
-        torch-version: ["2.5", "2.6"]
+        torch-version: ["2.5", "2.6", "2.7", "nightly"]
```
Comment (dayo09) [inline .github/workflows/check-pr.yaml:32]: NOTE: pytorch 2.7.0 release date is planed to 4/25, due to https://dev-discuss.pytorch.org/t/pytorch-release-2-7-0-call-for-features/2780.

File: test/modules/net/interpolate.py [hunk 2]
```diff
@@ -35,3 +38,18 @@ def forward(self, x):
 
     def get_example_inputs(self):
         return (torch.randn(1, 2, 3, 4),)
+
+
+@tag.skip_if(
+    not HAS_TORCH_OVER_28_DEV,
+    reason="This case is supported from torch 2.8.0.dev, yet.",
```
Comment (mhs4670go) [inline test/modules/net/interpolate.py:45]: Maybe the sentence could be like these.

"The case isn't supported yet — it will be from torch 2.8.0.dev."
"This case is supported starting from torch 2.8.0.dev."
"Support for this case has been added in torch.2.8.0.dev."

File: tico/passes/lower_to_resize_nearest_neighbor.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        graph_module = exported_program.graph_module
+        graph = graph_module.graph
+        for node in graph.nodes:
+            if not node.op == "call_function":
+                continue
+
+            if not (
+                (node.target == torch.ops.aten.index.Tensor)
+                or (node.target == torch.ops.aten.upsample_nearest2d.vec)
+            ):
```
Comment (mhs4670go) [inline tico/passes/lower_to_resize_nearest_neighbor.py:224]: (optional)
```suggestion
            if node.target not in [torch.ops.aten.index.Tensor, torch.ops.aten.upsample_nearest2d.vec]:
```

**Review by mhs4670go — DISMISSED** [review]
LGTM

**Comment by jinevening** [comment]
It seems that this PR includes some chanages about ResizeNearestNeighbor (not sure this is the only change), but they are hidden behind the commit message of "Upgrade torch nightly version".

I suggest to make an issue whenever we upgrade torch version, and list up related issues there. That will be helpful to track our decisions for supporting new torch versions.

@dayo09 Could you make an issue for this update?

**Comment by dayo09** [comment]
@jinevening Sorry that I forgot to link the issue #9.

File: tico/utils/validate_args_kwargs.py [hunk 4]
```diff
@@ -920,7 +920,7 @@ class SliceArgs:
     """
 
     input: torch.fx.Node
-    dim: int
+    dim: int = 0
```
Comment (dayo09) [inline tico/utils/validate_args_kwargs.py:923]: Some operators have void `dim` field in torch 2.8.dev

File: tico/passes/lower_to_resize_nearest_neighbor.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        [AFTER PASS]
+        input - aten.permute(NCHW_to_NHWC) - circle_custom.resize_nearest_neighbor - aten.permute(NHWC_to_NCHW) - output
+
+    Since torch 2.8, `torch.nn.functional.interpolate` is converted to aten.upsample_nearest2d.vec` op.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def convert_index_to_resize_nearest_neighbor(
```
Comment (dayo09) [inline tico/passes/lower_to_resize_nearest_neighbor.py:76]: This part is almost the same to the original code.

File: tico/passes/lower_to_resize_nearest_neighbor.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+            nhwc_to_nchw = graph.call_function(
+                torch.ops.aten.permute.default,
+                args=(resize_nearest_neighbor, [0, 3, 1, 2]),
+            )
+            # Not set meta for propagating replacing node's meta.
+            node.replace_all_uses_with(nhwc_to_nchw, propagate_meta=True)
+
+        return resize_nearest_neighbor
+
+    def convert_upsample_nearest2d_to_resize_nearest_neighbor(
```
Comment (dayo09) [inline tico/passes/lower_to_resize_nearest_neighbor.py:149]: This part is newly added.

File: test/modules/net/interpolate.py [hunk 7]
```diff
@@ -35,3 +38,18 @@ def forward(self, x):
 
     def get_example_inputs(self):
         return (torch.randn(1, 2, 3, 4),)
+
+
+@tag.skip_if(
+    not HAS_TORCH_OVER_28_DEV,
+    reason="The case isn't supported yet. It will be supported from torch 2.8.0.dev",
+)
+class InterpolateOnePointFive(torch.nn.Module):
```
Comment (dayo09) [inline test/modules/net/interpolate.py:47]: Q. Why it's not supported in lower version?

A. 
- CASE(torch < 2.8.0.dev) 
  - it's converted to index. -> It's interpreted to "Multiple indices" case. -> it's not supported in tico yet.
- CASE(torch >= 2.8.0.dev)  
  - it's converted to upsample nearest 2d.  -> it converts directly to RNN(ResizeNearsestNeighbor) in tico. -> Can support 1.5 times.

File: tico/passes/lower_to_resize_nearest_neighbor.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+                torch.ops.circle_custom.resize_nearest_neighbor,
+                args=(nchw_to_nhwc, [expected_H, expected_W]),
+            )
+            nhwc_to_nchw = graph.call_function(
+                torch.ops.aten.permute.default,
+                args=(resize_nearest_neighbor, [0, 3, 1, 2]),
+            )
+            # Not set meta for propagating replacing node's meta.
+            node.replace_all_uses_with(nhwc_to_nchw, propagate_meta=True)
+            return resize_nearest_neighbor
```
Comment (jinevening) [inline tico/passes/lower_to_resize_nearest_neighbor.py:209]:
```suggestion
        return resize_nearest_neighbor
```

File: test/modules/net/interpolate.py (hunk 7 above)
Comment (jinevening) [reply test/modules/net/interpolate.py:47]: Can I ask what is "Multiple indices" case?

File: test/modules/net/interpolate.py [hunk 9]
```diff
@@ -13,6 +13,9 @@
 # limitations under the License.
 
 import torch
+from tico.utils.utils import HAS_TORCH_OVER_28_DEV
+
+from test.utils import tag
 
 
 class InterpolateDouble(torch.nn.Module):
```
Comment (jinevening) [inline test/modules/net/interpolate.py:21]: (not in this PR) BTW, it seems that this file can be moved to `op`, not `net`.

**Comment by jinevening** [comment]
Ah, I missed #9 (added a back link to the main comment). I left some comments. PTAL

File: test/modules/net/interpolate.py (hunk 7 above)
Comment (dayo09) [reply test/modules/net/interpolate.py:47]: @jinevening `IndexTensorVisitor` don't support multiple indices.

```py
        # TODO Support multiple indices
        if len(indices) - indices.count(None) > 1:  # type: ignore[arg-type]
            raise NotYetSupportedError(
                "Multiple indices is not supported yet in aten.index.Tensor"
            )
```

This is current implementaion. I left it as TODO.
I didn't look into what values were. :thinking:

**Comment by dayo09** [comment]
#37 must be merged to resolve the lint error.

File: tico/utils/validate_args_kwargs.py [hunk 10]
```diff
@@ -852,7 +852,7 @@ class ReshapeArgs:
 @dataclass
 class ResizeNearestNeighborArgs:
     """
-    # Only consider `torch.nn.functional.interpolate(x, scale_factor=2.0, mode='nearest')` case.
+    # Maps from `torch.nn.functional.interpolate(x, scale_factor=scale_factor, mode='nearest')` case.
```
Comment (miusic) [inline tico/utils/validate_args_kwargs.py:855]:
```suggestion
    # Mapped from `torch.nn.functional.interpolate(x, scale_factor=scale_factor, mode='nearest')` case.
```
(nit)
May be..?

File: tico/passes/lower_to_resize_nearest_neighbor.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+            %_prop_tensor_constant0 : [num_users=1] = placeholder[target=_prop_tensor_constant0]
+            %_prop_tensor_constant1 : [num_users=1] = placeholder[target=_prop_tensor_constant1]
+            %x : [num_users=1] = placeholder[target=x]
+            %_to_copy : [num_users=1] = call_function[target=torch.ops.aten._to_copy.default](args = (%x,), kwargs = {dtype: torch.float32})
+            %index : [num_users=1] = call_function[target=torch.ops.aten.index.Tensor](args = (%_to_copy, [None, None, %_prop_tensor_constant1, %_prop_tensor_constant0]), kwargs = {})
+            %_to_copy_3 : [num_users=1] = call_function[target=torch.ops.aten._to_copy.default](args = (%index,), kwargs = {dtype: torch.float32})
+            return (_to_copy_3,)
+
+        [BEFORE PASS]
+        input - aten.index - output
```
Comment (miusic) [inline tico/passes/lower_to_resize_nearest_neighbor.py:65]:
```suggestion
        [BEFORE PASS]
        case 1.
        input - aten.index - output

        case 2.
        input - aten.upsample_nearest2d.vec - output
```

It would be nice to add a case for `aten.upsample_nearest2d.vec` :D

**Review by miusic — APPROVED** [review]
LGTM with some comments 👍

File: tico/passes/lower_to_resize_nearest_neighbor.py (hunk 11 above)
Comment (dayo09) [reply tico/passes/lower_to_resize_nearest_neighbor.py:65]: Let me apply in the next PR :-D

File: test/pt2_to_circle_test/test_pt2_to_circle.py [hunk 12]
```diff
@@ -71,7 +71,12 @@ def convert_nnmodule_to_pt2(
     model: torch.nn.Module, example_inputs: tuple, pt2_model_path: str
 ):
     # Create .pt2 model
-    with torch.no_grad(), SuppressWarning(UserWarning, ".*quantize_per_tensor"):
```
Comment (mhs4670go) [inline test/pt2_to_circle_test/test_pt2_to_circle.py:74]: Not related with this PR but seems that `quantize_per_tensor` warning has been removed.

**Review by mhs4670go — APPROVED** [review]
LGTM

**Review by jinevening — APPROVED** [review]
LGTM
